### PR TITLE
Vectorize IndexOfAnyExcept for four values

### PR DESCRIPTION
**Pull request by adamsitnik**
I took a look and the usage of `IndexOfAnyExcept ` and the patterns are following:
- `IndexOfAnyExcept(byte)` - most frequent
- `IndexOfAnyExcept(int)`, `IndexOfAnyExcept(char)` - sporadic
- `IndexOfAnyExcept(" \t\r\n")` - few places

Since the first two got vectorized by @stephentoub in #73488 I've vectorized the 4 values case.

Benchmarks:

```cs
public class IndexOfAnyExcept
{
    private string _whiteSpaces, _noWhiteSpaces;

    [Params(1, 4, 16, 64, 256, 1024)]
    public int Length { get; set; }

    [GlobalSetup]
    public void Setup()
    {
        _whiteSpaces = new string(' ', Length);
        _noWhiteSpaces = new string('a', Length);
    }

    [Benchmark]
    public int ImmediateMismatch() => _noWhiteSpaces.AsSpan().IndexOfAnyExcept(" \t\r\n");

    [Benchmark]
    public int NoMismatch() => _whiteSpaces.AsSpan().IndexOfAnyExcept(" \t\r\n");
}
```

```ini
BenchmarkDotNet=v0.13.1.1845-nightly, OS=Windows 11 (10.0.22000.795/21H2)
AMD Ryzen Threadripper PRO 3945WX 12-Cores, 1 CPU, 24 logical and 12 physical cores
.NET SDK=7.0.100-preview.7.22377.5
  [Host]     : .NET 7.0.0 (7.0.22.37506), X64 RyuJIT AVX2
  Job-OQFSEO : .NET 7.0.0 (42.42.42.42424), X64 RyuJIT AVX2
    Job-main : .NET 7.0.0 (42.42.42.42424), X64 RyuJIT AVX2
```

|            Method |      Job | Length |         Mean | Ratio |
|------------------ |--------- |------- |-------------:|------:|
| ImmediateMismatch |   Job-PR |      1 |     7.401 ns |  1.30 |
| ImmediateMismatch | Job-main |      1 |     5.679 ns |  1.00 |
|                   |          |        |              |       |
|        NoMismatch |   Job-PR |      1 |     7.259 ns |  1.32 |
|        NoMismatch | Job-main |      1 |     5.483 ns |  1.00 |
|                   |          |        |              |       |
| ImmediateMismatch |   Job-PR |      4 |     7.597 ns |  1.34 |
| ImmediateMismatch | Job-main |      4 |     5.689 ns |  1.00 |
|                   |          |        |              |       |
|        NoMismatch |   Job-PR |      4 |    10.021 ns |  0.76 |
|        NoMismatch | Job-main |      4 |    13.155 ns |  1.00 |
|                   |          |        |              |       |
| ImmediateMismatch |   Job-PR |     16 |     9.135 ns |  1.47 |
| ImmediateMismatch | Job-main |     16 |     6.226 ns |  1.00 |
|                   |          |        |              |       |
|        NoMismatch |   Job-PR |     16 |     9.973 ns |  0.24 |
|        NoMismatch | Job-main |     16 |    41.561 ns |  1.00 |
|                   |          |        |              |       |
| ImmediateMismatch |   Job-PR |     64 |     9.123 ns |  1.47 |
| ImmediateMismatch | Job-main |     64 |     6.216 ns |  1.00 |
|                   |          |        |              |       |
|        NoMismatch |   Job-PR |     64 |    15.135 ns |  0.09 |
|        NoMismatch | Job-main |     64 |   164.828 ns |  1.00 |
|                   |          |        |              |       |
| ImmediateMismatch |   Job-PR |    256 |     9.126 ns |  1.47 |
| ImmediateMismatch | Job-main |    256 |     6.196 ns |  1.00 |
|                   |          |        |              |       |
|        NoMismatch |   Job-PR |    256 |    31.836 ns |  0.05 |
|        NoMismatch | Job-main |    256 |   642.908 ns |  1.00 |
|                   |          |        |              |       |
| ImmediateMismatch |   Job-PR |   1024 |     9.128 ns |  1.48 |
| ImmediateMismatch | Job-main |   1024 |     6.153 ns |  1.00 |
|                   |          |        |              |       |
|        NoMismatch |   Job-PR |   1024 |   104.512 ns |  0.04 |
|        NoMismatch | Job-main |   1024 | 2,534.292 ns |  1.00 |

It's clearly a tradeoff:
- for all strings with immediate mismatch, the method has regressed by 2-3ns
- same goes for very short strings with no mismatch
- for other cases, when there is no mismatch the longer the string the bigger the gain

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs [hunk 1]
```diff
@@ -1270,5 +1270,66 @@ public static int SequenceCompareTo<T>(ref T first, int firstLength, ref T secon
...
+                    }
+                }
+
+                [MethodImpl(MethodImplOptions.AggressiveInlining)]
+                static int ComputeIndex(ref T searchSpace, ref T current, Vector128<T> notEquals)
+                {
+                    uint notEqualsElements = notEquals.ExtractMostSignificantBits();
+                    int index = BitOperations.TrailingZeroCount(notEqualsElements);
+                    return index + (int)(Unsafe.ByteOffset(ref searchSpace, ref current) / Unsafe.SizeOf<T>());
+                }
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs:1329]: Can this be shared with the other overload?

**Comment by danmoseley** [comment]
Some pretty nice improvements!

**Comment by adamsitnik** [comment]
The CI did not start... I am going to close and re-open the PR

File: src/libraries/System.Private.CoreLib/src/System/MemoryExtensions.cs [hunk 2]
```diff
@@ -619,6 +680,9 @@ public static ReadOnlyMemory<char> AsMemory(this string? text, Range range)
                 case 3:
                     return IndexOfAnyExcept(span, values[0], values[1], values[2]);
 
+                case 4: // common for searching whitespaces
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/MemoryExtensions.cs:683]: Nit: (don't restart CI just for this)
```suggestion
                case 4: // common for searching ASCII whitespaces
```

File: src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs [hunk 3]
```diff
@@ -1220,19 +1220,72 @@ public static int IndexOfAnyExcept<T>(ref T searchSpace, T value0, int length)
...
+                        return ComputeIndex(ref searchSpace, ref oneVectorAwayFromEnd, notEquals);
+                    }
                 }
             }
 
             return -1;
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static int ComputeIndex<T>(ref T searchSpace, ref T current, Vector128<T> notEquals) where T : struct, IEquatable<T>
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs:1282]: There are many other methods in this type, whereas this "ComputeIndex" is only relevant to LastIndexOfExcept.  Consider renaming it accordingly.

**Review by stephentoub — APPROVED** [review]
Thanks!

**Comment by adamsitnik** [comment]
@stephentoub Since I currently can't merge #73768, I am going to merge this PR now and address your feedback later (renaming `ComputeIndex` to `ComputeFirstIndex`). Normally I would do it first, but we have limited time ;)

**Pull request merged**